### DISCUSSION TITLE
Router.replace in useEffect creates infinite loop

### DISCUSSION THREAD

**Discussion by 28JayG** [vercel/next.js#46616] [Help] [answered]
I have some code in my useEffect which is conditionally redirects the app to  a new route. The issue I face is when the redirect occurs the useEffect goes into infinite loop. 
PS: setting reactStrictMode to false solves the issue, but I don't want to do that.

here's the code snippet:

`
  const useMoveToUpload = (category: CategoryType) => {
    const router = useRouter();
  
    const requiredDocuments = useStore((state) =>
      getRequiredDocuments(state, category)
    );
  
    useEffect(() => {
      if (!requiredDocuments) return;
  
      const docTypes = memoizedObjectKeys(requiredDocuments);
      if (docTypes.length > 1) return;
  
      const requiredDocument = requiredDocuments[docTypes[0] as DocumentType];
      if (requiredDocument.numRequired && requiredDocument.numRequired > 1)
        return;
  
      router.replace(
        `${AppRoutes.UPLOAD}?category=${category}&documentType=${docTypes[0]}`
      );
    }, [category, requiredDocuments]);
  };
`

**Comment by ciruz** [comment]
What about something like:

```tsx
const useMoveToUpload = (category: CategoryType) => {
const router = useRouter();

const requiredDocuments = useStore((state) =>
  getRequiredDocuments(state, category)
);

useEffect(() => {
  if (!requiredDocuments) return;

  const docTypes = memoizedObjectKeys(requiredDocuments);
  if (docTypes.length > 1) return;

  const requiredDocument = requiredDocuments[docTypes[0] as DocumentType];
  if (requiredDocument.numRequired && requiredDocument.numRequired > 1)
    return;

    const currentRoute = router.pathname;
    const uploadRoute = `${AppRoutes.UPLOAD}?category=${category}&documentType=${docTypes[0]}`;

    //check if the routes are not the same
    if (currentRoute !== uploadRoute) {
      router.replace(uploadRoute);
    }
}, [category, requiredDocuments]);
};
```

**Reply by 28JayG** [reply]
@ciruz I tried that didn't work, even though I see the upload url in the browser, the app is still in the useEffect

**Reply by icyJoseph** [reply]
Try with `router.asPath` instead of `router.pathname`.

**Reply by 28JayG** [reply]
same result

**Reply by ciruz** [reply]
@28JayG can you console.log the router + uploadRoute for us?

@icyJoseph is right, my mistake.

The ``pathname`` looks like this: ``/foobar/foobar``.
While the ``asPath`` would look like this ``/foobar/foobar?foo=bar``

**Reply by icyJoseph** [reply]
One more thing, could you check `router.isReady` within the useEffect, if its not ready, terminate early as well (and add it to the dependencies).

**Reply by 28JayG** [reply]
adding router to dependency array? will it be a good call?

**Reply by icyJoseph** [reply]
Only, `router.isReady` - as dependency. `[category, requiredDocuments, router.isReady]`

**Reply by 28JayG** [reply]
adding router.isReady didn't help
by the way log you asked for @ciruz 
`{
    "router": {
        "pathname": "/collect",
        "route": "/collect",
        "query": {
            "category": "residence"
        },
        "asPath": "/collect/?category=residence",
        "components": {
            "/missing-documents": {
                "initial": true,
                "props": {
                    "pageProps": {}
                }
            },
            "/_app": {
                "styleSheets": []
            },
            "/collect": {
                "styleSheets": [],
                "props": {
                    "pageProps": {}
                },
                "route": "/collect",
                "query": {
                    "category": "residence"
                },
                "resolvedAs": "/collect/?category=residence"
            }
        },
        "isFallback": false,
        "basePath": "",
        "isReady": true,
        "isPreview": false,
        "isLocaleDomain": false,
        "events": {}
    },
    "uploadRoute": "/upload?category=residence&documentType=bankStatement"
}`

**Reply by ciruz** [reply]
I shortend it:

```tsx
{
    router: {
      pathname: "/collect",
      route: "/collect",
      query: { category: "residence" },
      asPath: "/collect/?category=residence",
    },
    uploadRoute: "/upload?category=residence&documentType=bankStatement",
  };
```

The paths are totaly different. 

I would need some more code to help.

**Reply by 28JayG** [reply]
how could we go about that?

**Reply by 28JayG** [reply]
do you need the code where I call the hook?

**Comment by 28JayG** [comment, marked as answer]
Solved this issue wrapped the router.replace() effect with setInteval & clearing the interval in useEffect return. 
This ensures that our effect runs after remounting

> React intentionally remounts your components in development to help you find bugs. The right question isn’t “how to run an Effect once”, but “how to fix my Effect so that it works after remounting”.

> Usually, the answer is to implement the cleanup function. The cleanup function should stop or undo whatever the Effect was doing. The rule of thumb is that the user shouldn’t be able to distinguish between the Effect running once (as in production) and a setup → cleanup → setup sequence (as you’d see in development).
